File: src/leetcode/array/js/RandomPickWithWeight.js
/**
 * @param {number[]} w
 */
var Solution = function(w) {

    this.sum = [w[0]];
    for(let i = 1; i < w.length; i++) {
        this.sum.push(this.sum[i-1] + w[i]);
    }
    this.size = this.sum.length-1;
};

/**
 * @return {number}
 */
Solution.prototype.pickIndex = function() {
    const {sum, size} = this;
    const index = Math.floor(Math.random() * this.sum[size]+1);
    let left = 0;
    let right = size;

    while(left < right){
        const mid = (left + right) >> 1;
        if(sum[mid] == index)
            return mid;
        else if(sum[mid] < index)
            left = mid + 1;
        else
            right = mid;
    }
    return left;
};

/**
 * Your Solution object will be instantiated and called as such:
 * var obj = new Solution(w)
 * var param_1 = obj.pickIndex()
 */

File: src/leetcode/array/binarysearch/RandomPickWithWeight.java
package leetcode.array.binarysearch;

import java.util.Random;
/*
    528. Random Pick with Weight
    https://app.gitbook.com/@timmybeeflin/s/cracking-leetcode/binary-search/528.-random-pick-with-weight

    time: O(n)
    space: O(n)
 */
public class RandomPickWithWeight {
    Random random;
    int sum[];

    public RandomPickWithWeight(int[] w) {
        for (int i = 1; i < w.length; i++) {
            w[i] += w[i - 1];
        }
        sum = w;
        random = new Random();
    }

    public int pickIndex() {
        int index = random.nextInt(sum[sum.length - 1]) + 1;
        int left = 0;
        int right = sum.length - 1;
        while (left < right) {
            int mid = (left + right) >>> 1;

            if (sum[mid] == index) {
                return mid;
            } else if (sum[mid] < index) {
                left = mid + 1;
            } else {
                right = mid;
            }
        }
        return left;
    }

/**
 * Your Solution object will be instantiated and called as such:
 * Solution obj = new Solution(w);
 * int param_1 = obj.pickIndex();
 */
}
